### ISSUE TITLE
Could you review and provide feedback on a MutationObserver implementation example?

### ISSUE THREAD

**Issue by johnmaxwell** [turbolinks/turbolinks#159]
I've seen MutationObserver mentioned a number of times by @sstephenson in connection with Turbolinks. I was intrigued, because I have been trying to figure out how to do Turbolinks implementations The Right Way (TM). But I was not really familiar with MutationObserver, and I couldn't find a concrete example of its usage with Turbolinks, so I put together a little test Rails 5 app that uses MutationObserver and Turbolinks:

https://github.com/agileleague/rails-5-turbolinks-mutation-observer-example

It's basically a foothold for understanding how a Rails developer could use MutationObserver with Turbolinks The Right Way (TM). I imagine there is plenty of room for improvement. Would you mind reviewing it and leaving some feedback on the MO/TL implementation? I'll try to improve the repo using your feedback, and hopefully it can fill a hole in community documentation for MutationObserver and Turbolinks.

Thanks!

**Comment by ZempTime** [comment]
Chiming in to say I've been undergoing the exact same need, and would be happy to help flesh out details or proofread documentation around MutationObserver/Custom HTML elements. I know its not specifically in the purview of what Turbolinks is actually responsible for, but I've got some ugly code in my applications that these things are the solution for, and the code in question was specifically introduced by my choice to use Turbolinks. 

TL;DR - I would also find it useful to have some resources around this present/linked to in the readme, wherever they might live, and would be happy to help.

p.s. - Turbolinks maintainers - the readme is beautifully written. I've been gushing about it. Thanks for your effort, I deeply recognize and appreciate it.

**Comment by javan** [comment]
@johnmaxwell, your implementation is on the right track! A couple notes: 
* You'll want to check that a mutation's `addedNodes` are matched against your selector in addition to its children. `element.querySelectorAll(selector)` will only find child elements.
* You'll want to handle `attributes` `mutation.type`s separately from `childList` since the element of interest is `mutation.target`, which won't be found in `mutation.addedNodes`.

I put together an example (https://jsfiddle.net/javan/rxyd0tts/) that also demonstrates using custom elements (since @ZempTime asked).

**Comment by johnmaxwell** [comment]
@javan Thank you for this feedback! It took me a while, but I have revised my example project incorporating your help and some other lessons learned. If you have further feedback, I'd love to hear it.

https://github.com/agileleague/rails-5-turbolinks-mutation-observer-example

**Comment by fschwahn** [comment]
@javan have you thought about adding a simple example on how to ideally use `MutationObserver` with turbolinks to the README? The linked mdn article is a bit overwhelming for the uninitiated, I think an example would go a long way.

**Comment by johnmaxwell** [comment]
@fschwahn I agree that that could be helpful. I'm happy to help out in refining https://github.com/agileleague/rails-5-turbolinks-mutation-observer-example into a better example and community resource. (Anyone) please feel free to open tickets in that repo with suggestions.

**Comment by johnlpollard** [comment]
I love the example here! One question, how do you handle MutationObserver with turbolinks cache? The observer will fire that an element has been added once for the cache and then again for when the page updates.

**Comment by packagethief** [comment]
@johnlpollard, Turbolinks adds a `data-turbolinks-preview` attribute to the `<html>` element when it renders a preview from cache. You can check for this and ignore the mutation.

In Basecamp we use a little helper:

```coffee
BC.displayingTurbolinksPreview = ->
  document.documentElement.hasAttribute("data-turbolinks-preview")
```

This could be a nice addition to Turbolinks proper.

**Comment by johnlpollard** [comment]
@packagethief thanks for the helpful info! 

Any reason why a mutation observer class like [this](https://github.com/agileleague/rails-5-turbolinks-mutation-observer-example/blob/master/app/assets/javascripts/selector_observer.coffee), isn't included with Turbolinks?
